
Привет, меня зовут Головатский Николай, я буду проверять твой проект.

В Практикуме принято общаться на "ты", однако если это неудобно - дай знать, и можем перейти на более официальное общение.

Ниже ты найдешь мои комментарии - **пожалуйста, не перемещай, не изменяй и не удаляй их,** это может усложнить и затянуть повторную проверку. 

Я буду использовать цветовую разметку: 

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера✅:</b> Если все сделано хорошо.</div>

<div class="alert alert-block alert-warning">
<b>Комментарий ревьюера⚠️:</b> Некритичное замечание, реккомендация на будущее. Для принятия проекта исправление не обязательно.</div>

<div class="alert alert-block alert-danger">
<b>Комментарий ревьюера❌:</b> Критичная ошибка, необходимо исправить, иначе проект не будет принят.
</div>

Давай работать над проектом в диалоге: если ты что-то меняешь в проекте или отвечаешь на мои комменатри — пиши об этом. Мне будет легче отследить изменения, если ты выделишь свои комментарии:
Например так:
<div class="alert alert-block alert-info">
<b>Комментарий студента:</b> 
Hello World!
</div>

Если писать особо нечего, просто оставляй комментарий с галочкой ☑️.

<div class="alert alert-block alert-success">
<b>Общее впечатление:</b> 
    
- Спасибо за работу :)
- Определенно могу сказать, что проект очень сильный;
- Мне очень понравилось, как ты детально погружаешься в данные и не упускаешь никаких мелочей;
- В проекте осталось пара моментов, после правки которых, я смогу принять работу😉
    
</div>

# Исследование надежности заемщиков.

<div class="alert alert-block alert-danger">
<b>Комментарий ревьюера❌:</b> На этом этапе было бы неплохо описать суть проекта. Краткое введение делает проект понятней и помогает ориентироваться
</div>

На данных кредитного отдела банка мы проведем исследование влияния семейного положения и количества детей клиента на факт погашения кредита в срок.

**Цель исследования** — ответить на четыре вопроса:
1. Есть ли зависимость между количеством детей и возвратом кредита в срок?
2. Есть ли зависимость между семейным положением и возвратом кредита в срок?
3. Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
4. Как разные цели кредита влияют на его возврат в срок?

**Ход исследования**
Данные о клиентах мы возьмем из файла `/datasets/data.csv`. О качестве данных ничего не известно. Поэтому необходимо провести обзор данных. 
 
Таким образом, исследование пройдёт в три этапа:
 1. Обзор данных.
 2. Предобработка данных.
 3. Ответы на вопросы.


<div class="alert alert-block alert-success">
<b>Комментарий ревьюера_v2✅:</b> Огонь🔥🔥🔥
</div>

### Шаг 1. Обзор данных

In [1]:
import pandas as pd
pd = pd.read_csv('/datasets/data.csv')
pd.head(5)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

### Шаг 2.1 Заполнение пропусков

Проверяем какие столбцы не заполнены:

In [2]:
pd.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


<div class="alert alert-block alert-success">
<b>Комментарий ревьюера✅:</b> Хорошее начало, информация о данных изучена👍</div>

Не заполнены две колонки days_employed и total_income.<br>
Эти две колонки связаны с данными о работе, поэтому есть два варианта почему они отсутствуют:<br>
&emsp;1) Клиенты отказались заполнять данную информацию;<br>
&emsp;2) Данные не подтягиваются из сторонних систем.<br>
Перед тем как посчитать медианные значение необходимо исправить ошибки в колонке days_employed, выведем экстремумы положительных и отрицательных значений:

In [3]:
print('Максимальное положительное значение:', int(pd['days_employed'][pd['days_employed'] > 0 ].max()))
print('Минимальное положительное значение:', int(pd['days_employed'][pd['days_employed'] > 0 ].min()))
print('Максимальное отрицательное значение:', int(pd['days_employed'][pd['days_employed'] < 0 ].max()))
print('Минимальное отрицательное значение:', int(pd['days_employed'][pd['days_employed'] < 0 ].min()))

Максимальное положительное значение: 401755
Минимальное положительное значение: 328728
Максимальное отрицательное значение: -24
Минимальное отрицательное значение: -18388


<div class="alert alert-block alert-warning">
<b>Комментарий ревьюера⚠️:</b> Для изучения распределения количественного столбца подойдет метод describe() 

In [4]:
# Код ревьюера
pd.days_employed.describe()

count     19351.000000
mean      63046.497661
std      140827.311974
min      -18388.949901
25%       -2747.423625
50%       -1203.369529
75%        -291.095954
max      401755.400475
Name: days_employed, dtype: float64

Положительные значения имеют сильно больший порядок, чем отрицательные, делаем предположение, что нарушена размерность и указаны часы, а не дни, так как в днях сложно ориентироваться посмотрим значения в годах:

In [5]:
print('Максимальное количество лет стажа:', int(pd['days_employed'][pd['days_employed'] > 0 ].max() / 24 / 364))
print('Минимальное количество лет стажа:', int(pd['days_employed'][pd['days_employed'] > 0 ].min() / 24 / 364))

Максимальное количество лет стажа: 45
Минимальное количество лет стажа: 37


Наше предположение, о том что в положительных значениях указаны часы корректно.<br>
Необходимо выяснить на сколько корректно, что у данных клиентов слишком большой стаж 37 и 45 лет, путем проверки среднего возраста клиентов с положительными значениями:

In [6]:
print('Средний возраст клиентов:', int(pd['dob_years'][pd['days_employed'] > 0 ].mean()))

Средний возраст клиентов: 59


Средний возраст клиента 59 лет, для них стаж в 37-45 лет вполне нормально.<br>
Переводим часы в дни для положительных значений и берем модуль для отрицательных значений и выводим получившиеся значения:

In [7]:
pd.loc[(pd['days_employed'] > 0), 'days_employed'] = (pd['days_employed'] / 24)
pd.loc[(pd['days_employed'] < 0), 'days_employed'] = (abs(pd['days_employed']))
print('Максимальное положительное значение:', int(pd['days_employed'][pd['days_employed'] > 0 ].max()))
print('Минимальное положительное значение:', int(pd['days_employed'][pd['days_employed'] > 0 ].min()))

Максимальное положительное значение: 18388
Минимальное положительное значение: 24


Находим долю пропущенных значений в столбцах:

<div class="alert alert-block alert-warning">
<b>Комментарий ревьюера⚠️:</b> Ты обработал значения в столбце, но не учел, что он не участвует в исследовании, так что его можно было оставить в исходном состоянии) 

In [8]:
days_employed_missed = len(pd[pd['days_employed'].isna()]) / len(pd['days_employed'])
total_income_missed = len(pd[pd['total_income'].isna()]) / len(pd['total_income'])
print(f'Доля пропущенных значений в столбце days_employed: {days_employed_missed :.0%}')
print(f'Доля пропущенных значений в столбце total_income: {total_income_missed :.0%}')

Доля пропущенных значений в столбце days_employed: 10%
Доля пропущенных значений в столбце total_income: 10%


Отсутствуют 10% значений.<br>
Это большое количество отсутствующих значений, которое повлияет на результаты, соответственно их необходимо заполнить либо средними, либо медианными значениями.<br>
Обоснование выбора медианного значения для заполнения, а не среднего:

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера✅:</b> Долю пропусков можно посмотреть так - df.isna().mean() 😉</div>

In [9]:
days_employed_mean = pd['days_employed'].mean()
days_employed_median = pd['days_employed'].median()
total_income_mean = pd['total_income'].mean()
total_income_median = pd['total_income'].median()
print('Среднее значение по колонке days_employed:', int(days_employed_mean))
print('Медианное значение по колонке days_employed:', int(days_employed_median))
print('Среднее значение по колонке total_income:', int(total_income_mean))
print('Медианное значение по колонке total_income:', int(total_income_median))

Среднее значение по колонке days_employed: 4641
Медианное значение по колонке days_employed: 2194
Среднее значение по колонке total_income: 167422
Медианное значение по колонке total_income: 145017


Медианные значения меньше, по тому что в выборку не попадают сверх огромные зарплаты и экстремально низкие зарплаты, поэтому более соответствуют действительности.<br>
Заполняем отсутствующие значения медианными и убеждаемся, что колонки заполнены:

In [10]:
pd['days_employed'] = pd['days_employed'].fillna(value=days_employed_median)
pd['total_income'] = pd['total_income'].fillna(value=total_income_median)
pd.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       21525 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


<div class="alert alert-block alert-warning">
<b>Комментарий ревьюера⚠️:</b> Заполнить пропуски медианами - хорошее решение. 
    
Есть еще один, более интересный способ заполнения пропусков. Например, доход клиента (total_income) можно заполнить медианными значениями в разрезе income_type - для каждого income_type посчитать медианный total_income и заполнить пропуски ими. Так мы не усредним данные и меньше повлияем на распределение)

### Шаг 2.2 Проверка данных на аномалии и исправления.

#### Проверка значений колонки children:

In [11]:
pd['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

Есть некорректные значения -1 и 20, заменим их на среднее значение количества детей. <br>
Значение -1 это скорее всего отсутствие данных, а 20 это техническая ошибка системы. <br>
Расчет среднего количества детей на клиента:

In [12]:
corret_cn = [0,1,2,3,4,5]
corret_cn_df = pd[pd['children'].isin(corret_cn)]
print('Среднее количество детей:', corret_cn_df['children'].mean())

Среднее количество детей: 0.4731800766283525


Среднее количество детей 0.47, поэтому вставляем 0 на место не корректных значений и проводим проверку:

In [13]:
pd['children'] = pd['children'].replace(-1, 0)
pd['children'] = pd['children'].replace(20, 0)
pd['children'].value_counts()

0    14272
1     4818
2     2055
3      330
4       41
5        9
Name: children, dtype: int64

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера✅:</b> Логичное предположение, засчитано🔥</div>

#### Проверка значений колонки dob_years:

In [14]:
pd['dob_years'].value_counts()

35    617
40    609
41    607
34    603
38    598
42    597
33    581
39    573
31    560
36    555
44    547
29    545
30    540
48    538
37    537
50    514
43    513
32    510
49    508
28    503
45    497
27    493
56    487
52    484
47    480
54    479
46    475
58    461
57    460
53    459
51    448
59    444
55    443
26    408
60    377
25    357
61    355
62    352
63    269
64    265
24    264
23    254
65    194
66    183
22    183
67    167
21    111
0     101
68     99
69     85
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64

101 значение "0", заменяем отсутствующие значения на средние и проверяем результат:

In [15]:
pd.loc[(pd['dob_years'] == 0), 'dob_years'] = (int(pd['dob_years'][pd['dob_years'] != 0].mean()))
print('Количество нулевых значений:', len(pd[pd['dob_years'] == 0]))

Количество нулевых значений: 0


<div class="alert alert-block alert-success">
<b>Комментарий ревьюера✅:</b> Согласен😉</div>

#### Проверка значений колонки education:

In [16]:
pd['education'].value_counts()

среднее                13750
высшее                  4718
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   274
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
УЧЕНАЯ СТЕПЕНЬ             1
Ученая степень             1
Name: education, dtype: int64

Все значения корректные, но отличается регистр, соответственно поправляем регистр и проверяем результат:

In [17]:
pd['education'] = pd['education'].str.lower()
pd['education'].value_counts()

среднее                15233
высшее                  5260
неоконченное высшее      744
начальное                282
ученая степень             6
Name: education, dtype: int64

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера✅:</b> Все верно😉</div>

#### Проверка значений колонки education_id:

In [18]:
pd['education_id'].value_counts()

1    15233
0     5260
2      744
3      282
4        6
Name: education_id, dtype: int64

Проверяем что id уникален:

In [19]:
pd[['education', 'education_id']].drop_duplicates().reset_index(drop=True)

education  education_id
0               высшее             0
1              среднее             1
2  неоконченное высшее             2
3            начальное             3
4       ученая степень             4

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера✅:</b> 🔥🔥🔥</div>

 #### Проверка значений колонки family_status:

In [20]:
pd['family_status'].value_counts()

женат / замужем          12380
гражданский брак          4177
Не женат / не замужем     2813
в разводе                 1195
вдовец / вдова             960
Name: family_status, dtype: int64

Необходимо привести все значения к единому регистру и провести проверку.<br>
Такие ошибки возникают при ручном вводе данных оператором.

In [21]:
pd['family_status'] = pd['family_status'].str.lower()
pd['family_status'].value_counts()

женат / замужем          12380
гражданский брак          4177
не женат / не замужем     2813
в разводе                 1195
вдовец / вдова             960
Name: family_status, dtype: int64

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера✅:</b> Тут все хорошо🔥 </div>

#### Проверка значений колонки family_status_id:

In [22]:
pd['family_status_id'].value_counts()

0    12380
1     4177
4     2813
3     1195
2      960
Name: family_status_id, dtype: int64

Со значениями все в порядке, необходимо удостовериться, что id уникален:

In [23]:
pd[['family_status', 'family_status_id']].drop_duplicates().reset_index(drop=True)

family_status  family_status_id
0        женат / замужем                 0
1       гражданский брак                 1
2         вдовец / вдова                 2
3              в разводе                 3
4  не женат / не замужем                 4

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера✅:</b> 👍 </div>

#### Проверка значений колонки gender:

In [24]:
pd['gender'].value_counts()

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

Удаляем строку с одним отсутствующим значением, на расчеты не повлияет 

In [25]:
pd = pd.drop(pd[pd['gender'] == 'XNA'].index)

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера✅:</b> Согласен😉</div>

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера✅:</b> Тут что то не то с заголовком😉</div>

<div class="alert alert-block alert-info">
<b>Комментарий студента:</b> 
Убрал лишний заголовок)
</div>

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера_v2✅:</b> 👍
</div>

#### Проверка значений колонки income_type:

In [26]:
pd['income_type'].value_counts()

сотрудник          11119
компаньон           5084
пенсионер           3856
госслужащий         1459
предприниматель        2
безработный            2
в декрете              1
студент                1
Name: income_type, dtype: int64

Значения в колонке логически корректны.

<div class="alert alert-block alert-warning">
<b>Комментарий ревьюера⚠️:</b> Отлично, но я бы добавил вывод, что все в порядке. Еще ты задублил это пункт😉</div>

#### Проверка значений колонки debt:

In [27]:
pd['debt'].value_counts()

0    19783
1     1741
Name: debt, dtype: int64

Значения в колонке логически корректны.

#### Проверка колонки income_type на отсутствие значение <= 0

In [28]:
pd[pd['total_income'] <= 0]

Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose]
Index: []

#### Проверка значений колонки purpose:

In [29]:
pd['purpose'].value_counts()

свадьба                                   797
на проведение свадьбы                     777
сыграть свадьбу                           774
операции с недвижимостью                  676
покупка коммерческой недвижимости         664
операции с жильем                         653
покупка жилья для сдачи                   653
операции с коммерческой недвижимостью     651
жилье                                     647
покупка жилья                             647
покупка жилья для семьи                   641
строительство собственной недвижимости    635
недвижимость                              634
операции со своей недвижимостью           630
строительство жилой недвижимости          626
покупка недвижимости                      623
покупка своего жилья                      620
строительство недвижимости                620
ремонт жилью                              612
покупка жилой недвижимости                607
на покупку своего автомобиля              505
заняться высшим образованием      

Значения в колонке логически корректны.

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера✅:</b> Молодец, все аномалии изучены и исправлены верно. Двигаемся дальше🔥</div>

### Шаг 2.3. Изменение типов данных.

Заменяем вещественный тип данных в столбце total_income на целочисленный.
Выведем тип данных колонки total_income:

In [30]:
pd['total_income'].dtypes

dtype('float64')

Производим замену типа в колонке total_income и проверяем тип в результирующей колонки:

In [31]:
pd['total_income'] = pd['total_income'].astype('int64')
pd['total_income'].dtypes

dtype('int64')

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера✅:</b> На этом этапе нечего добавить, все круто👍</div>

### Шаг 2.4. Удаление дубликатов.

In [32]:
count_pd = len(pd)
pd = pd.drop_duplicates().reset_index(drop=True)
print('Количество удаленных дубликатов строк:',count_pd -len(pd))

Количество удаленных дубликатов строк: 71


<div class="alert alert-block alert-success">
<b>Комментарий ревьюера✅:</b> Молодец, дубликаты удалены верно👍

### Шаг 2.5. Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.

Создаем словарь по колонке education:

In [33]:
education_dict = pd[['education_id','education']].drop_duplicates().reset_index(drop=True)
education_dict

education_id            education
0             0               высшее
1             1              среднее
2             2  неоконченное высшее
3             3            начальное
4             4       ученая степень

Создаем словарь по колонке family_status:

In [34]:
family_status_dict = pd[['family_status_id','family_status']].drop_duplicates().reset_index(drop=True)
family_status_dict

family_status_id          family_status
0                 0        женат / замужем
1                 1       гражданский брак
2                 2         вдовец / вдова
3                 3              в разводе
4                 4  не женат / не замужем

Удаляем колонки отправленные в словари и проверяем результирующую таблицу:

In [35]:
pd = pd.drop(columns=['education', 'family_status'])
pd.head(5)

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1    8437.673028         42             0                 0      F   
1         1    4024.803754         36             1                 0      F   
2         0    5623.422610         33             1                 0      M   
3         3    4124.747207         32             1                 0      M   
4         0   14177.753002         53             1                 1      F   

  income_type  debt  total_income                     purpose  
0   сотрудник     0        253875               покупка жилья  
1   сотрудник     0        112080     приобретение автомобиля  
2   сотрудник     0        145885               покупка жилья  
3   сотрудник     0        267628  дополнительное образование  
4   пенсионер     0        158616             сыграть свадьбу

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера✅:</b> Сделано верно, молодец🔥</div>

### Шаг 2.6. Категоризация дохода.

Создаем функцию категоризации по определенным правилам, применяем ее к столбцу total_income_category и производим проверку:

In [36]:
def total_income_category(row):
    if 0 <= row['total_income'] <= 30000 :
        return 'E'
    if 30001 <= row['total_income'] <= 50000 :
        return 'D'
    if 50001 <= row['total_income'] <= 200000 :
        return 'C'
    if 200001 <= row['total_income'] <= 1000000 :
        return 'B'
    if row['total_income'] >= 1000001 :
        return 'A'
    return 'error'

pd['total_income_category'] = pd.apply(total_income_category, axis=1) 

pd['total_income_category'].value_counts()

C    16016
B     5040
D      350
A       25
E       22
Name: total_income_category, dtype: int64

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера✅:</b> Функция написана верно, круто что результат ее выполнения проверен👍

### Шаг 2.7. Категоризация целей кредита.

Данные заполнены от руки и есть основные 4 цели для взятия кредита, соответственно создаем функцию категоризации и применяем ее к столбцу total_income_category и производим проверку:

In [37]:
def age_group_unemployed(row):
    purpose = row['purpose']
    if 'свадьб' in purpose:
        return 'проведение свадьбы'
    if 'недвиж' in purpose or 'жиль' in purpose:
        return 'операции с недвижимостью'
    if 'авто' in purpose:
        return 'операции с автомобилем'
    if 'образ' in purpose:
        return 'получение образования'
    
pd['purpose_category'] = pd.apply(age_group_unemployed, axis=1)

pd['purpose_category'].value_counts()

операции с недвижимостью    10810
операции с автомобилем       4306
получение образования        4013
проведение свадьбы           2324
Name: purpose_category, dtype: int64

In [38]:
pd.head(5)

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1    8437.673028         42             0                 0      F   
1         1    4024.803754         36             1                 0      F   
2         0    5623.422610         33             1                 0      M   
3         3    4124.747207         32             1                 0      M   
4         0   14177.753002         53             1                 1      F   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0        253875               покупка жилья   
1   сотрудник     0        112080     приобретение автомобиля   
2   сотрудник     0        145885               покупка жилья   
3   сотрудник     0        267628  дополнительное образование   
4   пенсионер     0        158616             сыграть свадьбу   

  total_income_category          purpose_category  
0                     B  операции с недвижимостью  
1                     C    операции с автомобилем  
2                     C  операции с недвижимостью  
3                     B     получение образования  
4                     C        проведение свадьбы

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера✅:</b> Тут все верно, функция отрабатывает правильно. Проверка результата ее выполнений проведена👍</div>

### Шаг 2.8. Ответы на вопросы.

##### Вопрос 1: Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [39]:
first_quest = pd.pivot_table(index='children', columns='debt', values='dob_years', aggfunc='count')
first_quest['percent_off_debt'] = first_quest[1] / (first_quest[0] + first_quest[1]) * 100
first_quest

debt            0       1  percent_off_debt
children                                   
0         13141.0  1072.0          7.542391
1          4364.0   444.0          9.234609
2          1858.0   194.0          9.454191
3           303.0    27.0          8.181818
4            37.0     4.0          9.756098
5             9.0     NaN               NaN

##### Вывод 1:

Зависимость приблизительно 2% между критериями есть дети/нет детей, независимо от количества, так же необходимо учитывать, что расчет по количеству детей - 3, 4 и 5 необходимо использовать аккуратно при принятии решения, из-за не репрезентативности выборки.

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера✅:</b> Ответ верный👍</div>

##### Вопрос 2: Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [40]:
second_quest = pd.pivot_table(index='family_status_id', columns='debt', values='dob_years', aggfunc='count')
second_quest['percent_off_debt'] = second_quest[1] / (second_quest[0] + second_quest[1]) * 100
second_quest = second_quest.merge(family_status_dict, how='inner', on='family_status_id')[[0, 1, 'family_status', 'percent_off_debt']].sort_values(by=['percent_off_debt']).reset_index(drop=True)
second_quest

0    1          family_status  percent_off_debt
0    896   63         вдовец / вдова          6.569343
1   1110   85              в разводе          7.112971
2  11408  931        женат / замужем          7.545182
3   3762  388       гражданский брак          9.349398
4   2536  274  не женат / не замужем          9.750890

##### Вывод 2:

Минимальное количество просрочек допускают вдовцы/вдовы и люди в разводе - 6.6%, 7.1% соответственно;<br>
Чуть больше просрочек допускают люди в зарегестрированном браке - 7.5%;<br>
И с большим отрывом почти в 2% клиенты с не зарегестрированным браком и не женат / не замужем 9.3%, 9.8%.

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера✅:</b> Вывод правильный👍)</div>

##### Вопрос 3: Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [41]:
third_quest =  pd.pivot_table(index='total_income_category', columns='debt', values='dob_years', aggfunc='count')
third_quest['percent_off_debt'] = third_quest[1] / (third_quest[0] + third_quest[1]) * 100
third_quest

debt                       0     1  percent_off_debt
total_income_category                               
A                         23     2          8.000000
B                       4684   356          7.063492
C                      14656  1360          8.491508
D                        329    21          6.000000
E                         20     2          9.090909

##### Вывод 3:

Клиенты с очень низким и очень высоким доходом составляют не репрезентаитивныю выборку, поэтому их не рассматриваем, категории А, Е.<br>
Самое маленькое количество просрочек допускают люди с доходом от 30 тыс. руб до 50 тыс. руб. - 6%,<br>
Второе по величение количества просрочек допускают люди с доходом от 200 тыс. руб до 1 млн. руб. - 7.06%,<br>
Самое большое количество просрочек допускают люди с доходом от 50 тыс. руб до 200 тыс. руб. - 8.49%

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера✅:</b> Согласен😉</div>

##### Вопрос 4: Как разные цели кредита влияют на его возврат в срок?

In [42]:
fourth_quest = pd.pivot_table(index='purpose_category', columns='debt', values='dob_years', aggfunc='count')
fourth_quest['percent_off_debt'] = fourth_quest[1] / (fourth_quest[0] + fourth_quest[1]) * 100
fourth_quest = fourth_quest.sort_values(by=['percent_off_debt'])
fourth_quest

debt                          0    1  percent_off_debt
purpose_category                                      
операции с недвижимостью  10028  782          7.234043
проведение свадьбы         2138  186          8.003442
получение образования      3643  370          9.220035
операции с автомобилем     3903  403          9.359034

##### Вывод 4:

Наименьшее количество клиентов допускающих просрочку с целью кредита "операции с недвижимостью" - 7.23%,<br>
Далее идут клиенты с целью кредита "проведение свадьбы", количество клиентов допускающих просрочку - 8%,<br>
Далее идут клиенты с целью кредита "получение образования" и "операции с автомобилем", количество клиентов допускающих просрочку - 9.22% и 9.33% соответственно.

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера✅:</b> Аргументировано👍</div>

## Общий вывод:

**Шаг 1. Обзор данных.**<br>
Часть значений в колонках days_employed и total_income не заполнена.<br>

**Шаг 2.1 Заполнение пропусков.**<br>
1. Заполняем медианными значениями колонку days_employed, предварительно скорректировав значения;<br>
    1. Положительные значения делим на 24, так как значения занесли в часах;<br>
    2. Отрицательные значения переводим в положительные.<br>
2. total_income заполняем медианными значениями;<br>

**Шаг 2.2 Проверка данных на аномалии и исправления.**<br>
1. Корректируем колонку "children" - заменяем некорректные значения -1 и 20 на среднее значение количества детей;<br>
2. Корректируем колонку "dob_years" - заменяем значение "0" на средние;<br>
3. Корректируем колонку "education" - делаем по всей колонке единый регистр;<br>
4. Корректируем колонку "family_status" - делаем по всей колонке единый регистр;<br>
5. Корректируем колонку "gender" - удаление одной строки со значением "XNA".<br>

**Шаг 2.3. Изменение типов данных.**<br>
Заменяем вещественный тип данных в столбце total_income на целочисленный для удобства дальнейших расчетов.<br>

**Шаг 2.4. Удаление дубликатов.**<br>
Удалена 71 задублированная строка.<br>

**Шаг 2.5. Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.<br>**
Для удобства использования датафрейма были сформированы 2 словаря:<br>
* education_dict <br>
* family_status_dict <br>

Удалены  колонки pd['education'], pd['family_status'].

**Шаг 2.6. Категоризация дохода.**<br>
Присваиваем клиенту категорию в зависимости от дохода в столбец pd['total_income_category'] на основании диапазонов, указанных ниже: <br>
* 0–30000 — 'E';<br>
* 30001–50000 — 'D';<br>
* 50001–200000 — 'C';<br>
* 200001–1000000 — 'B';<br>
* 1000001 и выше — 'A'.<br>

**Шаг 2.7. Категоризация целей кредита.**<br>
Присваиваем клиенту одну из четырех категорий в зависимости от цели кредита в столбец pd['purpose_category'] согласно списку ниже: <br>
* операции с автомобилем;<br>
* операции с недвижимостью;<br>
* проведение свадьбы;<br>
* получение образования.<br>

**Шаг 2.8. Ответы на вопросы.**<br>
* Вопрос 1: Есть ли зависимость между количеством детей и возвратом кредита в срок?<br>
    - Зависимость приблизительно 2% между критериями есть дети/нет детей, независимо от количества<br>
* Вопрос 2: Есть ли зависимость между семейным положением и возвратом кредита в срок?<br>
    - Минимальное количество просрочек допускают вдовцы/вдовы и люди в разводе - 6.6%, 7.1% соответственно;<br>
    - Чуть больше просрочек допускают люди в зарегестрированном браке - 7.5%;<br>
    - И с большим отрывом почти в 2% клиенты с не зарегестрированным браком и не женат / не замужем 9.3%, 9.8%.<br>
* Вопрос 3: Есть ли зависимость между уровнем дохода и возвратом кредита в срок?<br>
    - Клиенты с очень низким и очень высоким доходом составляют не репрезентаитивныю выборку, поэтому их не рассматриваем, категории А, Е.<br>
    - Самое маленькое количество просрочек допускают люди с доходом от 30 тыс. руб до 50 тыс. руб. - 6%,<br>
    - Второе по величение количества просрочек допускают люди с доходом от 200 тыс. руб до 1 млн. руб. - 7.06%,<br>
    - Самое большое количество просрочек допускают люди с доходом от 50 тыс. руб до 200 тыс. руб. - 8.49%<br>
* Вопрос 4: Как разные цели кредита влияют на его возврат в срок?<br>
    - Наименьшее количество клиентов допускающих просрочку с целью кредита "операции с недвижимостью" - 7.23%,<br>
    - Далее идут клиенты с целью кредита "проведение свадьбы", количество клиентов допускающих просрочку - 8%,<br>
    - Далее идут клиенты с целью кредита "получение образования" и "операции с автомобилем", количество клиентов    допускающих просрочку - 9.22% и 9.33% соответственно.<br>

**Вывод:**
На основе проведенного анализа можно построить модель кредитного скоринга состоящую из четырех факторов со следующим подразделением:<br>
* Наличие детей (да/нет):<br>
    - Есть дети;<br>
    - Нет детей.<br>
* Семейное положение:<br>
    - Вдовец / вдова;<br>
    - В разводе, женат / замужем;<br>
    - Гражданский брак, не женат / не замужем;<br>
* Уровень дохода:<br>
    - 0 тыс. руб. - 30 тыс. руб.;<br>
    - 30 тыс. руб. - 50 тыс. руб.;<br>
    - 50 тыс. руб. - 200 тыс. руб.;<br>
    - 200 тыс. руб. - 1 млн. руб.;<br>
    - более 1 млн. руб.;<br>
* Цель кредита.<br>
    - операции с недвижимостью<br>
    - проведение свадьбы<br>
    - получение образования<br>
    - операции с автомобилем<br>

Идеальный клиент обладает следующим набором факторов:<br>
1. Наличие детей - да;<br>
2. Семейное положение - вдовец / вдова;<br>
3. Уровень дохода - от 30 тыс. руб до 50 тыс. руб;<br>
4. Цель кредита - операции с недвижимостью.<br>






<div class="alert alert-block alert-danger">
<b>Комментарий ревьюера❌:</b> В конце проекта принято кратко описывать все проделанные шаги и полученные результаты. Зачем это нужно - когда проект захочет посмотреть будущий работодатель, у него может не быть времени на подробный разбор кода. Вероятнее всего он бегло просмотрит код, но захочет изучить результат, который будет в общем выводе. Поэтому все же советую написать общий вывод)
</div>

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера_v2✅:</b> Отличный вывод, другое дело🔥🔥🔥
</div>

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера✅:</b>Мне очень понравилась твоя работа. Все рассуждения верные, очень логичные и осмысленные. Ничего лишнего, всё по полкам. Ты подробно разбираешься в деталях, для будущего аналитика это очень важная черта😊

Проект еще нужно немного доработать, а именно:

- добавить краткое описание проекта в самом начале
    
- поправить общий вывод
    
Остальные замечания оставляю на твое усмотрение😉

Жду твою работу на повторное ревью.
    
[Тут есть полезные приемы Markdown](https://paulradzkov.com/2014/markdown_cheatsheet/)

</div>

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера_v2✅:</b> Отличная работа, проект принимаю. Могу с уверенностью сказать - ты хорошо владеешь пройденным материалом. Желаю удачи в дальнейшей учебе🔥 </div>